### PR TITLE
[Perf] Cache the ModelMetadata for known type "object"

### LINKED ISSUE #4377: Optimize creation of root VDD
Sampling data from HelloWorldMvc

image

We're creating this VDD with model metadata for `object` as a sensible default. We can do something smarter here like caching it to avoid the lookup.

### PR DESCRIPTION
**Pull request by mnltejaswini**
Fixes #4377 

Scenario : https://github.com/aspnet/Performance/tree/dev/testapp/HelloWorldMvc
Before Optimization
-----------------------
image

After Optimization
----------------------
image

### CODE REVIEW

**Comment by rynowak** [comment]
I don't think the data included here is sufficient to conclude anything, 6ms of sampling data is a tiny amount. I'll start reviewing the changes and let's talk tomorrow about how to gather more data.

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadataProvider.cs [hunk 1]
```diff
@@ -71,6 +74,11 @@ public virtual ModelMetadata GetMetadataForType(Type modelType)
                 throw new ArgumentNullException(nameof(modelType));
             }
 
+            if (modelType == typeof(object) && _cachedMetadataForObjectType != null)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadataProvider.cs]: Why is the null check needed here?

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadataProvider.cs]: This at least needs a comment about why it's important

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/ViewDataDictionaryControllerPropertyActivator.cs [hunk 2]
```diff
@@ -49,7 +49,6 @@ private PropertyActivator<ControllerContext>[] GetPropertiesToActivate(Type type
 
         private ViewDataDictionary GetViewDataDictionary(ControllerContext context)
         {
-            var serviceProvider = context.HttpContext.RequestServices;
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/ViewDataDictionaryControllerPropertyActivator.cs]: :heart_eyes_cat:

**Comment by mnltejaswini** [comment]
:up::date:

**Comment by rynowak** [comment]
Is there updated perf data for this? Come by if you want help collecting data

**Comment by mnltejaswini** [comment]
@rynowak  I updated the profiles with more sampling data

**Comment by dougbu** [comment]
Not getting a warm feeling from looking just at the time spent activating `ViewDataDictionary` properties in controllers. From #4377 we know that activation accounts for ~1.8% of execution time and the above data shows we save ~20% of that. But I'd prefer a more direct comparison.

On the other hand, I suspect we get the metadata for `typeof(object)` enough other places that there may be a larger change in the overall time spent in`GetMetadataForType()` relative to the rest of the app.

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadataProvider.cs [hunk 3]
```diff
@@ -84,6 +93,13 @@ private ModelMetadataCacheEntry CreateCacheEntry(ModelMetadataIdentity key)
             return new ModelMetadataCacheEntry(metadata, details);
         }
 
+        private ModelMetadata GetMetadataForObjectType()
+        {
+            var key = ModelMetadataIdentity.ForType(typeof(object));
+            var entry = CreateCacheEntry(key);
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadataProvider.cs]: Suggest changing `GetMetadataForProperties()` to call `GetMetadataForType()`. Right now, the cache entry for `typeof(object)` may be created anyhow, duplicating this work and uselessly hitting the cache despite the shortcut above.

**Comment by dougbu** [comment]
:watch:

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadataProvider.cs (hunk 3 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadataProvider.cs]: I don't understand this feedback - how would returning the cached type-metadata for object be correct when we need to return a property-metadata?

Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadataProvider.cs]: `GetMetadataForProperties()` first gets `modelType`'s (the container's) metadata, duplicating most of `GetMetadataForType()`. My feedback is to maintain the consistency or just to use `GetMetadataForType()` for that part of `GetMetadataForProperties()`.

Comment (mnltejaswini) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadataProvider.cs]: GetMetadataForProperties() doesn't use the DefaultModelMetadata for the container's ModelType. It uses DefaultModelMetadataDetails that has less info that the Metadata. Wondering if we can reuse the Object Metadata if any of the properties are of type object?

Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadataProvider.cs]: > Wondering if we can reuse the Object Metadata if any of the properties are of type `object`?

No, property metadata is a completely different beast from type metadata. I'm wondering if we can save and reuse this `ModelMetadataCacheEntry` for when anything calls `GetMetadataForProperties(typeof(object))`. See [line 49](https://github.com/aspnet/Mvc/pull/4518/files#diff-c8fa65ecc8077ca10cd5f923c0082de2R49).

**Comment by mnltejaswini** [comment]
:up::date:

**Comment by rynowak** [comment]
Data looks much more convincing! Thanks :+1:

File: test/Microsoft.AspNetCore.Mvc.Core.Test/ModelBinding/Metadata/DefaultModelMetadataProviderTest.cs [hunk 4]
```diff
@@ -45,6 +45,24 @@ public void GetMetadataForType_Cached()
...
+            // Act
+            var metadata1 = Assert.IsType<DefaultModelMetadata>(provider.GetMetadataForType(typeof(object)));
+            var metadata2 = Assert.IsType<DefaultModelMetadata>(provider.GetMetadataForType(typeof(object)));
+
+            // Assert
+            Assert.Same(metadata1, metadata2);
+            Assert.Same(metadata1.Attributes, metadata2.Attributes);
+            Assert.Same(metadata1.BindingMetadata, metadata2.BindingMetadata);
+            Assert.Same(metadata1.DisplayMetadata, metadata2.DisplayMetadata);
+            Assert.Same(metadata1.ValidationMetadata, metadata2.ValidationMetadata);
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/ModelBinding/Metadata/DefaultModelMetadataProviderTest.cs]: Don't bother with the additional assertions. Nothing can be different between properties of `metadata1` and `metadata2` once you've confirmed they are the same.

In addition, the `Assert.IsType()` checks definitely don't belong w/in `// Act`. Why are they needed at all?

**Comment by dougbu** [comment]
:shipit: but with fewer assertions.

**Pull request merged**